src/draw/sw/lv_draw_sw_layer.c

```diff
@@ -43,54 +43,56 @@
 struct _lv_draw_layer_ctx_t * lv_draw_sw_layer_create(struct _lv_draw_ctx_t * draw_ctx, lv_draw_layer_ctx_t * layer_ctx,
                                                       lv_draw_layer_flags_t flags)
 {
-    if(LV_COLOR_SCREEN_TRANSP == 0 && (flags & LV_DRAW_LAYER_FLAG_HAS_ALPHA)) {
-        LV_LOG_WARN("Rendering this widget needs LV_COLOR_SCREEN_TRANSP 1");
-        return NULL;
-    }
+    do {
```

Comment:
Hmm just my 2 cents but:

What you are basically do with the `do` is introduce an inline function, which early returns on success, otherwise frees the memory and return the failure (null).

I would solve part of the issue as follow:

I would change this line:

https://github.com/lvgl/lvgl/blob/f5759351806d8b14537c0488842e3c7132c72fc9/src/draw/lv_draw_layer.c#L56-L59

to something like:

```c

 lv_draw_layer_ctx_t * init_layer_ctx = draw_ctx->layer_init(draw_ctx, layer_ctx, flags);
 if(init_layer_ctx == NULL) { 
     lv_mem_free(layer_ctx);
     return NULL; 
 } 
 return  init_layer_ctx;
```

And remove all mem_free calls from the layer_init function. This way, the allocation and deallocation stays inside the same function.

Reply:
Hi @Viatorus 
Thank you for your suggestion. 
From my perspective, it is reasonable and good! But to be honest, I don't have sufficient information to make such a change and I am not the right guy to make such a decision.

As you can see, 

```c
draw_ctx->layer_init(draw_ctx, layer_ctx, flags);
```
This code is an interface invoking, there is a promise between the interface caller(interface user) and the interface callee (interface implementation). This promise determines what the caller could do and what the callee should do.

Your suggestion is not only fixing this memory leakage issue but also changing this promise, i.e. it is an optimization and refactory of the interface.

As you can see, this memory leakage issue is a serious problem which involves almost everyone. We need a quick fix but not change the behaviours of interface callers and callees.

To make it clear:

- **This PR is to create a quick fix but not optimize or refractory** 
- The change this PR made should not change the behaviour of an existing interface.
- Unless @kisvegabor who designed this interface agrees, I don't have sufficient information to adopt your suggestion

On the other hand, I am only an external contributor, when I making a fix, I prefer to:

- Keep the change minimal and within the scope of the original code
- Fix the issue without crossing interface boundaries

**My fix doesn't stop you to create a separate PR to introduce your optimisation solution later.** 

Please allow me to keep the current solution and wait for the opinion of @kisvegabor.

Reply:
>  lv_draw_layer_ctx_t * init_layer_ctx = draw_ctx->layer_init(draw_ctx, layer_ctx, flags);
 if(init_layer_ctx == NULL) { 
     lv_mem_free(layer_ctx);
     return NULL; 
 } 
 return  init_layer_ctx;
And remove all mem_free calls from the layer_init function. This way, the allocation and deallocation stays inside the same function.

This solution seems the best to me. Having `lv_mem_free(layer_ctx)` in `lv_draw_sw_layer_create` seems like a mistake to me as `layer_ctx` is not allocated by `lv_draw_sw_layer_create` so it shouldn't free it either.

Reply:
OK, with the interface author's confirmation. I will take your suggestion @Viatorus .